src/ProxyClient/Request/RequestQueue.php

```diff
@@ -0,0 +1,212 @@
+    
+    public function __construct(
+        array $servers,
+        $baseUrl = null
+    ) {
+        $this->setServers($servers);
+        $this->setBaseUrl($baseUrl);
+    }
+    
+    public function add(InvalidationRequest $request)
+    {
+        $signature = $request->getSignature();
+        
+        if (!isset($this->queue[$signature])) {
+            $this->queue[$signature] = $request;
+        }
+    }
+    
+    public function clear()
+    {
+        $this->queue = array();
+    }
+    
+    public function count()
+    {
+        return count($this->queue);
+    }
+
+    /**
+     * @return RequestInterface[]
+     */
+    public function all()
+    {
+        $requests = array();
+        foreach ($this->queue as $queuedRequest) {
+            $headers = $queuedRequest->getHeaders();
+            if (empty($headers['Host'])) {
+                if ($this->defaultHost) {
+                    $headers['Host'] = $this->defaultHost;
+                } else {
+                    unset($headers['Host']);
+                }
+            }
+
+            foreach ($this->servers as $server) {
+                $request = new Request(
+                    $queuedRequest->getMethod(),
+                    $this->combineUrls($server, $queuedRequest->getUrl()),
+                    $headers
+                );
+                
+                $requests[] = $request;
+            }
+        }
+        
+        return $requests;
+    }
+    
+    
+    /**
+     * Set caching proxy servers
+     *
+     * @param array $servers Caching proxy proxy server hostnames or IP
+     *                       addresses, including port if not port 80.
+     *                       E.g. array('127.0.0.1:6081')
+     *
+     * @throws InvalidUrlException If server is invalid or contains URL
+     *                             parts other than scheme, host, port
+     */
+    public function setServers(array $servers)
+    {
+        $this->servers = array();
+        foreach ($servers as $server) {
+            $this->servers[] = $this->filterUrl(
+                $server,
+                array('scheme', 'host', 'port')
+            );
+        }
+    }
+    
+    /**
+     * Set application hostname, optionally including a base URL, for purge and
+     * refresh requests
+     *
+     * @param string $url Your application’s base URL or hostname
+     */
+    public function setBaseUrl($url = null)
+    {
+        if (null === $url) {
+            $this->baseUrl = null;
+
+            return;
+        }
+            
+        $this->baseUrl = $this->filterUrl($url);
+        $parts = parse_url($this->baseUrl);
+        $this->defaultHost = $parts['host'];
+        if (isset($parts['port'])) {
+            $this->defaultHost .= ':' . $parts['port'];
+        }
+
+        if (isset($parts['path'])) {
+            $this->basePath = $parts['path'];
+        }
+    }
+    
+    public function getAllowedSchemes()
+    {
+        return array('http');
```

Comment:
yeah, lets just make protocol required and not guess neither validate it. most likely varnish invalidation should go directly to varnish and not via the ssl terminator, but if somebody really needs that i don't see why we should forbid it.